### DISCUSSION TITLE
How to implement the equivalent of `alt` for an arbitrary number of choices?

### DISCUSSION THREAD

**Discussion by beneyal** [winnow-rs/winnow#509] [Q&A] [answered]
Hello!

As part of my research, I'm porting some Scala code that uses the `atto` library (for incremental parsing), and I'm missing the `choice` combinator. Naturally, I reached for `winnow`'s `alt` combinator, but it (seems) to only work if the number of choices is known at compile time, either a tuple or a `[P; N]`. In my case, I have a `Vec<String>`,  I sort it by the length in descending order, and then wish to create a parser that will match the first case-insensitive string in the `Vec`.

At first, I tried something like the code below. Keep in mind that I'm still new to Rust, and this whole porting thing is also practice for me, so the code might not be idiomatic or performant.

```rust
fn choice<'a>(
    input: &'a mut Stream<'a>,
    choices: &'a [String],
) -> PResult<String, InputError<Stream<'a>>> {
    for c in choices {
        let mut p = Caseless(&c[..]);
        match p.parse_next(input).map(|s| s.to_owned()) {
            r @ Ok(_) => {
                return r;
            }
            Err(_) => {}
        }
    }

    fail.parse_next(input)
}
```

But that doesn't work because I discard partial parses. Then I tried to go back to my tried-and-true FP style, and use `alt` by folding right over the choices, with `fail` being the `init` value:

```rust
let p = choices.iter().rfold(fail, |acc, c| alt((acc, Caseless(&c[..]))));
```

This fails to compile with `expected fn item, found opaque type`. Trying something similar with a `for`  loop gives a similar error, i.e., this code:

```rust
fn choice<'a, E: ParserError<Stream<'a>>>(
    input: &'a mut Stream<'a>,
    choices: &'a [String],
) -> PResult<String, E> {
    let [fst, snd, rest @ ..] = choices else {
        let output = Caseless(&choices[0][..]).parse_next(input)?;
        return Ok(output.to_owned());
    };
    let mut fst = Caseless(&fst[..]);
    let mut snd = Caseless(&snd[..]);
    let mut p = alt((fst, snd));
    for c in choices.iter().skip(2) {
        let mut next = Caseless(&c[..]);
        p = alt((p, next));
    }
    p.map(|s| s.to_owned()).parse_next(input)
}
```
fails to mutate `p` inside the loop because `expected Caseless<&str>, found opaque type`.

So I'm lost 😅 

Any pointers are very appreciated! And feel free to tell me my code is terrible, I haven't worked with Rust enough to know, myself. 😉 Thanks in advance! ❤️

**Comment by epage** [comment, marked as answer]
In the tutorial we [build up `alt` from scratch](https://docs.rs/winnow/latest/winnow/_tutorial/chapter_3/index.html#alternatives).  Like the middle solution, you likely want to use `opt` to check if the parse succeeded or not.

**Reply by beneyal** [reply]
Thanks for the pointer! I came up with this:

```rust
fn choice<'i, E: ParserError<Stream<'i>>>(
    choices: &'i [String],
) -> impl Parser<Stream<'i>, String, E> {
    move |input: &mut Stream<'i>| {
        for choice in choices.iter() {
            if (opt(Caseless(&choice[..])).parse_next(input)?).is_some() {
                return Ok(choice.to_owned());
            }
        }
        fail.parse_next(input)
    }
}
```

and it  seems to work well! 😃
